File: client/src/models/payment-request.js
export class PaymentRequest {
  amount;
  from_address;
  to_label;
  to_address;
  approval_url = '';
  completed;

  user;
  user_id;

  listing;
  listing_id;

  constructor(data) {
    Object.assign(this, data);
    if (this._id) {
      this.id = this.getId();
    }
  }
  getId() {
    return this._id.$oid;
  }
}

File: app/models/payment_request.rb
class PaymentRequest
  include Mongoid::Document
  include Mongoid::Timestamps

  field :amount, type: String
  field :approval_url, type: String
  field :completed, type: Boolean, default: false

  belongs_to :user
  belongs_to :listing

  def self.with_user
    PaymentRequest.includes(:user, :listing)
  end

  def complete
    self.completed = true
    self.save!

    self.listing.complete
  end
end
